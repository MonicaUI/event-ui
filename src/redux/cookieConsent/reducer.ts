// Reducer to update state for cookie Consent
import { DENY } from '../../utils/CookieConstants';

export const SET_COOKIE_CONSENT = 'event-guestside-site/cookieConsent/SET_COOKIE_CONSENT';

export default function cookieConsentReducer(state = { status: DENY }, action: $TSFixMe): $TSFixMe {
  switch (action.type) {
    case SET_COOKIE_CONSENT:
      return {
        ...state,
        status: action.payload.cookieStatus
      };
    default:
      return state;
  }
}
