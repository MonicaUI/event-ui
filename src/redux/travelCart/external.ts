import Logger from '@cvent/nucleus-logging';
import { getIn } from 'icepick';
import {
  travelCartUpdated,
  clearUserSessionData,
  updateUnsavedCreditCard,
  removeUnSavedTravelAnswers
} from './actions';
import { SET_SELECTED_AIR_REQUEST_ID, SET_SELECTED_AIR_ACTUAL_ID, SET_SELECTED_GROUP_FLIGHT_ID } from './actionTypes';
import { updateTravelCartAndUpdateInState, persistTravelCartAndUpdateState, updateTravelCreditCard } from './internal';
import { getRegCart } from '../selectors/shared';
import { loadAirportsForTravelCart } from '../airports';
import { getPrimaryRegistrationId, getAttendeeId, getEventRegistrations } from '../registrationForm/regCart/selectors';
import { loadAirportsForEventAndTravelCart } from '../airports';
import { updateRegistrantInformationInTravelBookings, hasAnyTravelItem } from '../../utils/travelUtils';
import { getTravelCartSnapshotVersion } from './selectors';
import { loadTravelSnapshotVersion } from 'event-widgets/redux/modules/eventTravel';
import { updateTravelAnswers } from './travelAnswers';

const LOG = new Logger('redux/travelCart/external');

/**
 * Restores the travel cart from the backend into the state
 * @param {*} travelCartId The travel cart id corresponding to which the travel cart should be loaded
 */
export function restoreTravelCartIntoState(travelCartId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { travelApiClient }
    } = getState();
    try {
      const travelCart = await travelApiClient.getTravelCart(travelCartId);
      await dispatch(travelCartUpdated(travelCart));
      await dispatch(clearUserSessionData());
      await dispatch(loadTravelSnapshotVersion(getTravelCartSnapshotVersion(getState())));
      await dispatch(loadAirportsForEventAndTravelCart());
      LOG.debug(`restore travel cart with id ${travelCartId} was successful`);
    } catch (error) {
      LOG.error(`restore travel cart with id ${travelCartId} failed`, error);
      throw error;
    }
  };
}

/**
 * Persists the current travel cart in the state to backend
 * Used for making sure that the travel cart stays alive during the registration
 */
function persistTravelCart(travelForm, saveTravelCreditCard) {
  return async (dispatch, getState) => {
    const { cart, isCartCreated } = travelForm;
    // persist travel credit card
    let updatedCart = await dispatch(updateTravelCreditCard(cart, saveTravelCreditCard));
    // persist travelAnswers
    updatedCart = await dispatch(updateTravelAnswers(updatedCart));
    await persistTravelCartAndUpdateState(updatedCart, isCartCreated, dispatch, getState());
    await dispatch(updateUnsavedCreditCard(null)); // Clear credit card from userSession object
    await dispatch(removeUnSavedTravelAnswers()); // Clear travel answers from userSession object
  };
}

/**
 * Update reg type and admission item ids in travel cart corresponding to the values in reg cart
 * if anything has changed
 */
export function updateRegTypeAndAdmissionItemIdsInTravelBookings() {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      clients: { travelApiClient },
      travelCart: { cart, isCartCreated }
    } = state;
    if (!isCartCreated) {
      return;
    }
    const regCart = getRegCart(state);
    const modifiedResult = updateRegistrantInformationInTravelBookings(cart.bookings, regCart);
    if (modifiedResult.hasAnyBookingChanged) {
      await updateTravelCartAndUpdateInState(
        travelApiClient,
        {
          ...cart,
          bookings: modifiedResult.bookings
        },
        dispatch
      );
    }
  };
}

/**
 * Restores the transient travel cart corresponding to the regcart in state, if applicable
 */
export function restoreTransientTravelCartIntoState(forceLoadTransientCart?: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      clients: { travelApiClient },
      event,
      defaultUserSession: { isPlanner, isTestMode }
    } = state;
    const regCart = getRegCart(state);

    const travelConfigs = getIn(event, ['eventFeatureSetup', 'travelAndHousing']);
    if (
      ((regCart && !regCart.regCartId) || forceLoadTransientCart) &&
      travelConfigs &&
      (travelConfigs.hotelAccomodations || getIn(travelConfigs, ['airTravel', 'enabled']) || travelConfigs.carRental)
    ) {
      let bookingIdOverrides = {};
      const eventRegistrations = getEventRegistrations(regCart);
      Object.keys(eventRegistrations).forEach(
        eventRegId =>
          (bookingIdOverrides = {
            ...bookingIdOverrides,
            [getAttendeeId(regCart, eventRegId)]: eventRegId
          })
      );
      try {
        const response = await travelApiClient.getTransientTravelCart(
          getAttendeeId(regCart, getPrimaryRegistrationId(regCart)),
          true,
          { isForPlanner: isPlanner, isForTestMode: isTestMode, travelBookingIdOverrides: bookingIdOverrides }
        );
        // if response is successful set travelCart into state
        if (response.travelCart) {
          await dispatch(travelCartUpdated(response.travelCart));
          dispatch(loadAirportsForTravelCart());
        }
      } catch (error) {
        if (!(error.responseStatus && error.responseStatus === 404)) {
          throw error;
        }
      }
    }
  };
}

const requireSaving = (regCart, travelForm) => {
  return hasAnyTravelItem(travelForm) && !(regCart.regCancel && travelForm.cart.status === 'TRANSIENT');
};

/**
 * saves the current travel registration form to backend if needed
 * @param saveTravelCreditCard
 * @returns {Function}
 */
export function saveTravelRegistration(saveTravelCreditCard: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      travelCart: travelForm,
      registrationForm: { regCart }
    } = getState();
    if (requireSaving(regCart, travelForm)) {
      await dispatch(persistTravelCart(travelForm, saveTravelCreditCard));
    }
  };
}

/**
 * Retrieves a TravelCart from service when restoring the registration form
 * @param regCart
 * @returns {Function}
 */
export const restoreTravelRegistration = (regCart: $TSFixMe) => {
  return async (dispatch: $TSFixMe): Promise<$TSFixMe> => {
    // when invitee starts reg cancel, we load the transient cart, so can again load transient cart
    if (regCart.regCancel) {
      await dispatch(restoreTransientTravelCartIntoState(true));
    } else if (regCart.hasTravel) {
      await dispatch(restoreTravelCartIntoState(regCart.regCartId));
    }
  };
};

/**
 * start cancellation flow for travel
 * @returns {Function}
 */
export const startTravelCancellation = () => {
  return async (dispatch: $TSFixMe): Promise<$TSFixMe> => {
    /*
     * The actual cancel travel cart has requestStatus props for all bookings set to CANCEL, from that
     * we won't be able to tell whether the bookings were already cancelled or are coming for getting cancelled.
     * Therefore we're using the transient travel cart, where those props would be accurate for our purpose.
     * In case later we need any data specific to travel cancellation, we'd require loading the actual cancel cart then.
     */
    dispatch(restoreTransientTravelCartIntoState(true));
  };
};

/**
 * To add value to selectedGroupFlightIds in state
 */
export function setSelectedGroupFlightId(selectedGroupFlightId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedGroupFlightIds = [...getState().travelCart.userSession.groupFlights.selectedGroupFlightIds];
    if (!selectedGroupFlightIds.includes(selectedGroupFlightId)) {
      selectedGroupFlightIds.push(selectedGroupFlightId);
      dispatch({
        type: SET_SELECTED_GROUP_FLIGHT_ID,
        payload: {
          selectedGroupFlightIds
        }
      });
    }
  };
}

/**
 * To remove value from selectedGroupFlightIds in state
 * @param {Arrar<string>} idsToRemove
 */
export function removeSelectedGroupFlightIds(idsToRemove = []) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedGroupFlightIds = getState().travelCart.userSession.groupFlights.selectedGroupFlightIds || [];

    if (selectedGroupFlightIds.length > 0 && idsToRemove.length > 0) {
      const filteredIds = selectedGroupFlightIds.filter(existingId => {
        if (!idsToRemove.includes(existingId)) {
          return existingId;
        }
      });
      dispatch({
        type: SET_SELECTED_GROUP_FLIGHT_ID,
        payload: {
          selectedGroupFlightIds: filteredIds
        }
      });
    }
  };
}

/**
 * To add value to selectedAirRequestIds in state
 */
export function setSelectedAirRequestId(selectedAirRequestId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirRequestIds = [...getState().travelCart.userSession.airRequest.selectedAirRequestIds];
    if (!selectedAirRequestIds.includes(selectedAirRequestId)) {
      selectedAirRequestIds.push(selectedAirRequestId);
      dispatch({
        type: SET_SELECTED_AIR_REQUEST_ID,
        payload: {
          selectedAirRequestIds
        }
      });
    }
  };
}

/**
 * To remove value from selectedAirRequestIds in state
 * @param {Arrar<string>} idsToRemove
 */
export function removeSelectedAirRequestIds(idsToRemove = []) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirRequestIds = getState().travelCart.userSession.airRequest.selectedAirRequestIds || [];

    if (selectedAirRequestIds.length > 0 && idsToRemove.length > 0) {
      const filteredIds = selectedAirRequestIds.filter(existingId => {
        if (!idsToRemove.includes(existingId)) {
          return existingId;
        }
      });
      dispatch({
        type: SET_SELECTED_AIR_REQUEST_ID,
        payload: {
          selectedAirRequestIds: filteredIds
        }
      });
    }
  };
}

/**
 * To add value to selectedAirActualIds in state
 */
export function setSelectedAirActualId(selectedAirActualId: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirActualIds = [...getState().travelCart.userSession.airActual.selectedAirActualIds];
    if (!selectedAirActualIds.includes(selectedAirActualId)) {
      selectedAirActualIds.push(selectedAirActualId);
      dispatch({
        type: SET_SELECTED_AIR_ACTUAL_ID,
        payload: {
          selectedAirActualIds
        }
      });
    }
  };
}

/**
 * To remove value from selectedAirActualIds in state
 * @param {Arrar<string>} idsToRemove
 */
export function removeSelectedAirActualIds(idsToRemove = []) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const selectedAirActualIds = getState().travelCart.userSession.airActual.selectedAirActualIds || [];

    if (selectedAirActualIds.length > 0 && idsToRemove.length > 0) {
      const filteredIds = selectedAirActualIds.filter(existingId => {
        if (!idsToRemove.includes(existingId)) {
          return existingId;
        }
      });
      dispatch({
        type: SET_SELECTED_AIR_ACTUAL_ID,
        payload: {
          selectedAirActualIds: filteredIds
        }
      });
    }
  };
}
