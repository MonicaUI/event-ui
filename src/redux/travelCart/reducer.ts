import { LOG_OUT_REGISTRANT_SUCCESS } from '../registrantLogin/actionTypes';
import {
  INITIATE_CANCEL_REGISTRATION_SUCCESS,
  CREATE_REG_CART_SUCCESS,
  RESTORE_PARTIAL_CART_SUCCESS,
  CREATE_PLACEHOLDER_REG_CART_SUCCESS
} from '../registrationForm/regCart/actionTypes';
import {
  RESTORE_TRAVEL_CART,
  UPDATE_TRAVEL_CART_SUCCESS,
  TRAVEL_CART_NOT_FOUND,
  SET_SELECTED_AIR_ACTUAL_ID,
  SET_SELECTED_AIR_REQUEST_ID,
  SET_AIR_REQUEST_SUMMARY_VIEW,
  SET_AIR_ACTUAL_SUMMARY_VIEW,
  CLEAR_USER_SESSION_DATA,
  TOGGLE_OWN_AIR_TRAVEL_RESERVATION,
  TOGGLE_OWN_HOTEL_ACCOMMODATION,
  UPDATE_CREDIT_CARD,
  SET_EXPANDED_HOTELS,
  SET_ANOTHER_HOTEL_REQUEST,
  SET_HOTEL_REQUEST_SUMMARY_VIEW,
  RESET_USER_SESSION_TRAVEL_DATA,
  SEARCH_ROOMMATE_SUCCESS,
  SET_ROOMMATE_DATA,
  SET_TRAVEL_ANSWER_FIELD,
  REMOVE_TRAVEL_ANSWERS,
  SET_GROUP_FLIGHT_SUMMARY_VIEW,
  SET_SELECTED_GROUP_FLIGHT_ID,
  UPDATE_TRAVEL_ANSWERS
} from './actionTypes';
import { setIn } from 'icepick';

const initialState = {
  cart: {
    bookings: []
  },
  isCartCreated: false,
  roommates: {
    requestedFrom: [], // roommates who have already selected current registrant as roommate
    availableRoommates: {
      // other registrants/invitees that the current registrant can select as a roommate
      all: [], // all roommates returned by searches
      keywordSearch: {} // map of search by search terms
    }
  },
  userSession: {
    airRequest: { ownBooking: false, showSummary: true, selectedAirRequestIds: [] },
    hotelRequest: { ownBooking: {}, isSummaryView: false, expandedHotels: [], isMakingAnotherRequest: false },
    creditCard: null,
    airActual: { showSummary: true, selectedAirActualIds: [] },
    groupFlights: { showSummary: true, selectedGroupFlightIds: [] },
    travelAnswers: {}
  }
};

// eslint-disable-next-line complexity
export default function reducer(state = initialState, action: $TSFixMe): $TSFixMe {
  const { type, payload } = action;

  switch (type) {
    case RESTORE_TRAVEL_CART:
      return { ...payload } || { ...initialState };
    case UPDATE_TRAVEL_CART_SUCCESS:
      return {
        ...state,
        cart: {
          ...payload.travelCart
        },
        isCartCreated: true
      };
    case TRAVEL_CART_NOT_FOUND:
      return {
        ...state,
        cart: {
          id: payload.travelCartId
        },
        isCartCreated: false
      };
    case SET_ROOMMATE_DATA:
      return {
        ...state,
        roommates: {
          ...state.roommates,
          requestedFrom: payload.requestedFrom,
          availableRoommates: {
            ...state.roommates.availableRoommates,
            all: payload.availableRoommates,
            keywordSearch: {}
          }
        }
      };
    case SEARCH_ROOMMATE_SUCCESS:
      return {
        ...state,
        roommates: {
          ...state.roommates,
          availableRoommates: {
            ...state.roommates.availableRoommates,
            keywordSearch: {
              ...state.roommates.availableRoommates.keywordSearch,
              [payload.searchText]: payload.roommates
            }
          }
        }
      };
    case SET_SELECTED_AIR_REQUEST_ID:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          airRequest: {
            ...state.userSession.airRequest,
            selectedAirRequestIds: payload.selectedAirRequestIds
          }
        }
      };
    case SET_SELECTED_AIR_ACTUAL_ID:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          airActual: {
            ...state.userSession.airActual,
            selectedAirActualIds: payload.selectedAirActualIds
          }
        }
      };
    case SET_SELECTED_GROUP_FLIGHT_ID:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          groupFlights: {
            ...state.userSession.groupFlights,
            selectedGroupFlightIds: payload.selectedGroupFlightIds
          }
        }
      };
    case SET_AIR_REQUEST_SUMMARY_VIEW:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          airRequest: {
            ...state.userSession.airRequest,
            showSummary: payload.showSummary
          }
        }
      };
    case SET_AIR_ACTUAL_SUMMARY_VIEW:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          airActual: {
            ...state.userSession.airActual,
            showSummary: payload.showSummary
          }
        }
      };
    case SET_GROUP_FLIGHT_SUMMARY_VIEW:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          groupFlights: {
            ...state.userSession.groupFlights,
            showSummary: payload.showSummary
          }
        }
      };
    case CLEAR_USER_SESSION_DATA:
      return {
        ...state,
        userSession: { ...initialState.userSession }
      };
    case TOGGLE_OWN_AIR_TRAVEL_RESERVATION: {
      const userSession = !state.userSession.airRequest.ownBooking
        ? {
            ...state.userSession,
            airRequest: {
              ...state.userSession.airRequest,
              ownBooking: true,
              showSummary: true
            }
          }
        : {
            ...state.userSession,
            airRequest: {
              ...state.userSession.airRequest,
              ownBooking: false
            }
          };
      return {
        ...state,
        userSession
      };
    }
    case TOGGLE_OWN_HOTEL_ACCOMMODATION:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          // reset hotel request view to hotel request
          hotelRequest: {
            ...state.userSession.hotelRequest,
            ownBooking: {
              ...state.userSession.hotelRequest.ownBooking,
              [payload.currentRegistrationId]: !state.userSession.hotelRequest.ownBooking[payload.currentRegistrationId]
            },
            isSummaryView: false,
            expandedHotels: [],
            isMakingAnotherRequest: false
          }
        }
      };
    case SET_HOTEL_REQUEST_SUMMARY_VIEW:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          hotelRequest: {
            ...state.userSession.hotelRequest,
            isSummaryView: payload.showSummary
          }
        }
      };
    case SET_EXPANDED_HOTELS:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          hotelRequest: {
            ...state.userSession.hotelRequest,
            expandedHotels: payload.expandedHotels
          }
        }
      };
    case SET_ANOTHER_HOTEL_REQUEST:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          hotelRequest: {
            ...state.userSession.hotelRequest,
            isMakingAnotherRequest: payload.isMakingAnotherRequest
          }
        }
      };
    case RESET_USER_SESSION_TRAVEL_DATA:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          hotelRequest: {
            ...state.userSession.hotelRequest,
            isSummaryView: false,
            expandedHotels: [],
            isMakingAnotherRequest: false
          },
          airRequest: {
            ...state.userSession.airRequest,
            showSummary: true
          },
          airActual: {
            ...state.userSession.airActual,
            showSummary: true
          },
          groupFlights: {
            ...state.userSession.groupFlights,
            showSummary: true
          }
        }
      };
    case LOG_OUT_REGISTRANT_SUCCESS: // invitee logs out
    case INITIATE_CANCEL_REGISTRATION_SUCCESS: // invitee cancels reg
    case CREATE_REG_CART_SUCCESS: // When a fresh reg cart is created
    case CREATE_PLACEHOLDER_REG_CART_SUCCESS: // When an embedded reg temp cart is created
    case RESTORE_PARTIAL_CART_SUCCESS: // partial cart is restored
      return initialState;
    case UPDATE_CREDIT_CARD:
      return {
        ...state,
        userSession: {
          ...state.userSession,
          creditCard: payload.creditCard
            ? {
                ...payload.creditCard
              }
            : null
        }
      };
    case SET_TRAVEL_ANSWER_FIELD: {
      const { path, value, eventRegistrationId } = action.payload;
      return setIn(state, path, { ...value, eventRegistrationId });
    }
    case REMOVE_TRAVEL_ANSWERS: {
      return {
        ...state,
        userSession: {
          ...state.userSession,
          travelAnswers: {}
        }
      };
    }
    case UPDATE_TRAVEL_ANSWERS: {
      return {
        ...state,
        userSession: {
          ...state.userSession,
          travelAnswers: payload.updatedAnswers
        }
      };
    }
    default:
      return state;
  }
}
