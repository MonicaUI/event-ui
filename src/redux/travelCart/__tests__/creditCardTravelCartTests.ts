import reducer from '../index';
import registrationFormReducer from '../../registrationForm/reducer';
import regCartStatusReducer from '../../regCartStatus';
import { saveTravelRegistration } from '../external';
// import { getUpdatedTravelBookingsForAirRequests, removeAirRequests } from '../airRequest';
import timezones from './fixtures/timezonesFixture.json';
import getStoreForTest from 'event-widgets/utils/testUtils';
import travelCartFixture from './fixtures/creditCardTravelCartFixtures.json';
import eventTravelReducer from 'event-widgets/redux/modules/eventTravel';

jest.mock('../../selectors/currentRegistrationPath', () => {
  return {
    isGuestProductSelectionEnabledOnRegPath: jest.fn(() => true)
  };
});

jest.mock('../../website/selectors', () => {
  return {
    ...jest.requireActual<$TSFixMe>('../../website/selectors'),
    __esModule: true,
    isWidgetPresentOnCurrentPage: jest.fn(() => true)
  };
});

jest.mock('../../reducer', () => {
  return {
    getCurrentPageId: jest.fn(() => true)
  };
});

jest.mock('../../../utils/travelUtils', () => {
  return {
    updateRegistrantInformationInTravelBookings: jest.fn(c => {
      return { bookings: c };
    }),
    hasAnyTravelItem: jest.fn(() => true)
  };
});

test('Save credit card in empty travel cart', async () => {
  const createTravelCartMock = jest.fn(cart => Promise.resolve({ travelCart: cart }));
  const updateRegCartMock = jest.fn((token, cart) => Promise.resolve({ regCart: cart }));
  const createCreditCardMock = jest.fn((token, card) =>
    Promise.resolve({
      contextId: 'dummyContextId',
      creditCard: card.creditCard
    })
  );

  const mockReducer = (state, action) => {
    return {
      accessToken: 'DUMMY_ACCESS_TOKEN',
      userSession: {},
      defaultUserSession: {
        eventId: 'EVENT_ID',
        isPreview: false,
        isPlanner: false
      },
      appData: {
        registrationSettings: {
          travelQuestions: []
        }
      },
      clients: {
        travelApiClient: {
          createTravelCart: createTravelCartMock,
          updateTravelCart: createTravelCartMock
        },
        regCartClient: {
          updateRegCart: updateRegCartMock
        },
        creditCardClient: {
          createCreditCard: createCreditCardMock
        }
      },
      eventTravel: eventTravelReducer(state.eventTravel, action),
      event: {
        isInTestMode: false,
        timezone: 35
      },
      registrationForm: registrationFormReducer(state.registrationForm, action),
      regCartStatus: regCartStatusReducer(state.regCartStatus, action),
      travelCart: reducer(state.travelCart, action),
      website: {
        theme: {
          global: {},
          sections: {}
        }
      },
      eventSnapshotVersion: 'EVENT_SNAPSHOT_VERSION',
      timezones
    };
  };
  const testRegCart = {
    regMod: false,
    regCartId: 'DUMMY_CART_ID',
    eventRegistrations: {
      DUMMY_BOOKING_ID: {
        registrationTypeId: 'b8e50dbf-7437-4eec-a964-bc27d79d0372',
        registrationPathId: 'REG_PATH_1',
        productRegistrations: [
          {
            productId: '3349dc98-e303-4bf9-b698-e1bdae85f946',
            productType: 'AdmissionItem',
            requestedAction: 'REGISTER'
          }
        ]
      },
      DUMMY_BOOKING_ID_TWO: {
        registrationTypeId: 'b8e50dbf-7437-4eec-a964-bc27d79d0373',
        registrationPathId: 'REG_PATH_2',
        productRegistrations: [
          {
            productId: '3349dc98-e303-4bf9-b698-e1bdae85f947',
            productType: 'AdmissionItem',
            requestedAction: 'REGISTER'
          }
        ]
      }
    }
  };
  const store = getStoreForTest(mockReducer, {
    registrationForm: {
      regCart: testRegCart,
      regCartPayment: {}
    },
    regCartStatus: { lastSavedRegCart: testRegCart },
    eventTravel: {
      travelSnapshotVersion: 'TRAVEL_SNAPSHOT_VERSION',
      airData: {}
    },
    travelCart: {
      cart: {
        bookings: []
      },
      isCartCreated: false,
      userSession: {
        airRequest: { ownBooking: false, showSummary: false, selectedAirRequestIds: [] },
        creditCard: {
          ...travelCartFixture.creditCard
        },
        travelAnswers: {}
      }
    }
  });

  await store.dispatch(saveTravelRegistration(true));
  expect(store.getState()).toMatchSnapshot();
  expect(createCreditCardMock).toBeCalledWith('DUMMY_ACCESS_TOKEN', {
    creditCard: {
      ...travelCartFixture.creditCardForCreateRequest
    }
  });
  expect(createTravelCartMock).toBeCalledWith({
    ...travelCartFixture.travelCartWithCreditCard,
    bookings: [
      {
        airBookings: [],
        hotelRoomBookings: [],
        airActuals: [],
        creditCard: {
          contextId: 'dummyContextId',
          creditCardDetails: {
            ...travelCartFixture.creditCard,
            cardType: undefined
          }
        },
        id: undefined,
        travelAnswers: [],
        concurAirActuals: [],
        pnrAirActuals: []
      }
    ],
    eventId: 'EVENT_ID'
  });
});
