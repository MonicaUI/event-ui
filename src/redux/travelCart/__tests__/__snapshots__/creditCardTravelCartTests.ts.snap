// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Save credit card in empty travel cart 1`] = `
Object {
  "accessToken": "DUMMY_ACCESS_TOKEN",
  "appData": Object {
    "registrationSettings": Object {
      "travelQuestions": Array [],
    },
  },
  "clients": Object {
    "creditCardClient": Object {
      "createCreditCard": [MockFunction] {
        "calls": Array [
          Array [
            "DUMMY_ACCESS_TOKEN",
            Object {
              "creditCard": Object {
                "accountHolderName": "dummy",
                "expMonth": "10",
                "expYear": "2018",
                "last4Digits": "3456",
                "number": "123456",
              },
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
    },
    "regCartClient": Object {
      "updateRegCart": [MockFunction],
    },
    "travelApiClient": Object {
      "createTravelCart": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "bookings": Array [
                Object {
                  "airActuals": Array [],
                  "airBookings": Array [],
                  "concurAirActuals": Array [],
                  "creditCard": Object {
                    "contextId": "dummyContextId",
                    "creditCardDetails": Object {
                      "accountHolderName": "dummy",
                      "cardType": undefined,
                      "expiryMonth": "10",
                      "expiryYear": "2018",
                      "last4Digits": "3456",
                      "number": "123456",
                    },
                  },
                  "hotelRoomBookings": Array [],
                  "id": undefined,
                  "pnrAirActuals": Array [],
                  "travelAnswers": Array [],
                },
              ],
              "eventId": "EVENT_ID",
              "eventSnapshotVersion": "EVENT_SNAPSHOT_VERSION",
              "id": "DUMMY_CART_ID",
              "isForPlanner": false,
              "isForPreview": false,
              "isForTestMode": false,
              "isPlannerApprovalRequired": false,
              "requestedAction": "BOOK",
              "travelSnapshotVersion": "TRAVEL_SNAPSHOT_VERSION",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "updateTravelCart": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "bookings": Array [
                Object {
                  "airActuals": Array [],
                  "airBookings": Array [],
                  "concurAirActuals": Array [],
                  "creditCard": Object {
                    "contextId": "dummyContextId",
                    "creditCardDetails": Object {
                      "accountHolderName": "dummy",
                      "cardType": undefined,
                      "expiryMonth": "10",
                      "expiryYear": "2018",
                      "last4Digits": "3456",
                      "number": "123456",
                    },
                  },
                  "hotelRoomBookings": Array [],
                  "id": undefined,
                  "pnrAirActuals": Array [],
                  "travelAnswers": Array [],
                },
              ],
              "eventId": "EVENT_ID",
              "eventSnapshotVersion": "EVENT_SNAPSHOT_VERSION",
              "id": "DUMMY_CART_ID",
              "isForPlanner": false,
              "isForPreview": false,
              "isForTestMode": false,
              "isPlannerApprovalRequired": false,
              "requestedAction": "BOOK",
              "travelSnapshotVersion": "TRAVEL_SNAPSHOT_VERSION",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
    },
  },
  "defaultUserSession": Object {
    "eventId": "EVENT_ID",
    "isPlanner": false,
    "isPreview": false,
  },
  "event": Object {
    "isInTestMode": false,
    "timezone": 35,
  },
  "eventSnapshotVersion": "EVENT_SNAPSHOT_VERSION",
  "eventTravel": Object {
    "airData": Object {},
    "travelSnapshotVersion": "TRAVEL_SNAPSHOT_VERSION",
  },
  "regCartStatus": Object {
    "lastSavedRegCart": Object {
      "eventRegistrations": Object {
        "DUMMY_BOOKING_ID": Object {
          "productRegistrations": Array [
            Object {
              "productId": "3349dc98-e303-4bf9-b698-e1bdae85f946",
              "productType": "AdmissionItem",
              "requestedAction": "REGISTER",
            },
          ],
          "registrationPathId": "REG_PATH_1",
          "registrationTypeId": "b8e50dbf-7437-4eec-a964-bc27d79d0372",
        },
        "DUMMY_BOOKING_ID_TWO": Object {
          "productRegistrations": Array [
            Object {
              "productId": "3349dc98-e303-4bf9-b698-e1bdae85f947",
              "productType": "AdmissionItem",
              "requestedAction": "REGISTER",
            },
          ],
          "registrationPathId": "REG_PATH_2",
          "registrationTypeId": "b8e50dbf-7437-4eec-a964-bc27d79d0373",
        },
      },
      "regCartId": "DUMMY_CART_ID",
      "regMod": false,
    },
  },
  "registrationForm": Object {
    "currentEventRegistrationId": "",
    "currentGuestEventRegistration": Object {},
    "discountCodeStatus": "",
    "errors": Object {},
    "preventRegistration": false,
    "regCart": Object {
      "eventRegistrations": Object {
        "DUMMY_BOOKING_ID": Object {
          "productRegistrations": Array [
            Object {
              "productId": "3349dc98-e303-4bf9-b698-e1bdae85f946",
              "productType": "AdmissionItem",
              "requestedAction": "REGISTER",
            },
          ],
          "registrationPathId": "REG_PATH_1",
          "registrationTypeId": "b8e50dbf-7437-4eec-a964-bc27d79d0372",
        },
        "DUMMY_BOOKING_ID_TWO": Object {
          "productRegistrations": Array [
            Object {
              "productId": "3349dc98-e303-4bf9-b698-e1bdae85f947",
              "productType": "AdmissionItem",
              "requestedAction": "REGISTER",
            },
          ],
          "registrationPathId": "REG_PATH_2",
          "registrationTypeId": "b8e50dbf-7437-4eec-a964-bc27d79d0373",
        },
      },
      "regCartId": "DUMMY_CART_ID",
      "regMod": false,
    },
    "regCartPayment": Object {},
    "validationMessages": Array [],
  },
  "timezones": Object {
    "35": Object {
      "abbreviation": "ET",
      "abbreviationResourceKey": "Event_Timezone_Abbr_35__resx",
      "dstInfo": Array [],
      "hasDst": true,
      "id": 35,
      "name": "Eastern Time",
      "nameResourceKey": "Event_Timezone_Name_35__resx",
      "plannerDisplayName": "(GMT-05:00) Eastern [US & Canada]",
      "utcOffset": -300,
    },
  },
  "travelCart": Object {
    "cart": Object {
      "bookings": Array [
        Object {
          "airActuals": Array [],
          "airBookings": Array [],
          "concurAirActuals": Array [],
          "creditCard": Object {
            "contextId": "dummyContextId",
            "creditCardDetails": Object {
              "accountHolderName": "dummy",
              "cardType": undefined,
              "expiryMonth": "10",
              "expiryYear": "2018",
              "last4Digits": "3456",
              "number": "123456",
            },
          },
          "hotelRoomBookings": Array [],
          "id": undefined,
          "pnrAirActuals": Array [],
          "travelAnswers": Array [],
        },
      ],
      "eventId": "EVENT_ID",
      "eventSnapshotVersion": "EVENT_SNAPSHOT_VERSION",
      "id": "DUMMY_CART_ID",
      "isForPlanner": false,
      "isForPreview": false,
      "isForTestMode": false,
      "isPlannerApprovalRequired": false,
      "requestedAction": "BOOK",
      "travelSnapshotVersion": "TRAVEL_SNAPSHOT_VERSION",
    },
    "isCartCreated": true,
    "userSession": Object {
      "airRequest": Object {
        "ownBooking": false,
        "selectedAirRequestIds": Array [],
        "showSummary": false,
      },
      "creditCard": null,
      "travelAnswers": Object {},
    },
  },
  "userSession": Object {},
  "website": Object {
    "theme": Object {
      "global": Object {},
      "sections": Object {},
    },
  },
}
`;
